### DISCUSSION TITLE
Seeing black background only for some images / sometimes

### DISCUSSION THREAD

**Discussion by ganymedio** [flbulgarelli/headbreaker#37] [Q&A] [answered]
Hi I'm seeing this behavior sometimes, where instead of the image I see only black. 

Any idea why it's happening and how to have more consistent performance? It's when I upload an image and try to use it in a puzzle.

Here's a screen capture video:

https://www.loom.com/share/fa5d17503c3f49cdacf5db4b0fadd2b1

And here's my index.html code

`      
<script src="scripts/headbreaker.js"></script>
      
      <h1>Upload Your Puzzle Image</h1>
      <input type="file" id="image-input" accept="image/jpeg, image/png, image/jpg">
      <div id="my-canvas"></div>
      <div id="autogen-canvas"></div>
    
    <script>
      const image_input = document.querySelector("#image-input");
      image_input.addEventListener("change", function() {
        const reader = new FileReader();
        reader.onload = function () {
          const img = new Image();
          img.src = reader.result;
          document.body.appendChild(img);
    
          const autogen = new headbreaker.Canvas('autogen-canvas', {
            width: 800, height: 650,
            pieceSize: 100, proximity: 20,
            borderFill: 10, strokeWidth: 1.5,
            lineSoftness: 0.18, image: img,
          });
    
          autogen.adjustImagesToPuzzleHeight();
          autogen.autogenerate({
            horizontalPiecesCount: 6,
            verticalPiecesCount: 5
          });
          autogen.draw();
        
          const canvas = new headbreaker.Canvas('my-canvas', {
            width: 800, height: 800, image: img
          });
          canvas.autogenerate();
          canvas.shuffle(0.7);
          canvas.draw();
          
        }
        reader.readAsDataURL(this.files[0]);
      });
    
    
    </script>
`

**Comment by flbulgarelli** [comment]
Hi! At fist glance your code looks right. I am wondering that may be related with the fact that you are reusing the same html `canvas` without releasing its resources. Have you tried - for instance - creating a new html `div` each time you load a new image and disposing the original one and all its children? If as you said it happens consistently with just one image I am wondering if there is something related with its aspect-ratio. 

Could you provide a full code pen, please?

**Comment by ganymedio** [comment]
Sure, thank you so much. I removed one of the canvases to simplify it, and tried removing the old div & adding a new one. 

https://codepen.io/bitcoinappdev/pen/yLEgBOW

With the way the code is now, I'm getting an error on image upload attempt, saying "Cannot read properties of null" so I guess I'm doing the creation of a new `div` and removal of an old one incorrectly?
 
![image](https://user-images.githubusercontent.com/17599867/200641688-9b47ceb9-fb9a-4d8c-9c77-f6d3c6d72727.png)

I'll keep looking into it but any more hints are appreciated, thanks!

**Comment by flbulgarelli** [comment]
Hi! I have checked your code and I managed to make it work. I found two issues with it: 

1. The code was adding a `canvas` element directly, instead of a `div` that acts as a the canvas holder
2. Also, although the code was reacting to the `load` event of the `reader` object, it was not waiting to the image to be properly loaded after its contents are set. In other words, an additional `img.onload` was missing. 

With those changes, it is now working: 

[Screencast from 27-11-22 15:28:49.webm](https://user-images.githubusercontent.com/677436/204153117-fdb0fd2a-2ba5-4820-80d3-17cc6f21cc41.webm)

**Comment by flbulgarelli** [comment, marked as answer]
PS: Here there is a minimal code sample based on your codepen that illustrates the modifications needed. Hope it helps!

```html
<script src="https://flbulgarelli.github.io/headbreaker/js/headbreaker.js"></script>
<body>
  <h1>Upload Your Puzzle Image</h1>
  <input type="file" id="image-input" accept="image/jpeg, image/png, image/jpg">
  <div id="puzzle-container">
  </div>
  <script>
    const puzzleContainer = document.getElementById('puzzle-container');
    const imageInput = document.getElementById("image-input");

    function loadImage(file, callback) {
      const reader = new FileReader();
      reader.readAsDataURL(file);
      reader.onload = function () {
        const image = new Image();
        image.src = reader.result;
        image.onload = () => callback(image);
      }
    }

    imageInput.addEventListener("change", function() {
      loadImage(this.files[0], (image) => {
        const oldPuzzle = document.getElementById('puzzle');
          if (oldPuzzle) {
            puzzleContainer.removeChild(oldPuzzle)
          }

          const newPuzzle = document.createElement('div');
          newPuzzle.id = 'puzzle';
          puzzleContainer.appendChild(newPuzzle);

          const autogen = new headbreaker.Canvas('puzzle', {
            width: 800, height: 650,
            pieceSize: 100, proximity: 20,
            borderFill: 10, strokeWidth: 1.5,
            lineSoftness: 0.18, image: image,
          });

          autogen.adjustImagesToPuzzleHeight();
          autogen.autogenerate({
            horizontalPiecesCount: 4,
            verticalPiecesCount: 4
          });
          autogen.draw();
      })
    });
  </script>
</body>
```

**Comment by ganymedio** [comment]
Wow, thank you so much, I appreciate your help!

There's no way to do animated gifs as puzzles with Headbreaker, is there? My gifs to render as static images for the puzzle.

**Comment by flbulgarelli** [comment]
Hi @LyteNetworkIO! Yes, you actually can animate puzzle images - although there is no explicit, documented support for it. However, it is a little hacky since you can not use animated `gift`s or `apng`s - you will need to handle the animation frames on your own and update the image buffers. 

In order to keep questions organized, I am now moving to questions to [Github Discussions](https://github.com/flbulgarelli/headbreaker/discussions). You you need more help with animations, please open a new question there. Thanks!
